### ISSUE TITLE
Add ownership behavior with `checkout` and `checkin` callbacks

### ISSUE THREAD

**Issue by fishcakez** [elixir-ecto/db_connection#149]
After #148 we need a way to handle changes that occur on checkout and checkin so we can still create Ecto sandbox. This needs to be able to change the module and state, then reverse it before returning to the pool (on checkin, disconnect and stop). It should take an options argument so we know whether to start transaction or not in sandbox.

**Comment by josevalim** [comment]
Closing in favor of #156.
